File: Python/minMoves2.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: minMoves2.py
@Function: 最少移动次数使数组元素相等 II 数学问题
@Link: https://leetcode-cn.com/problems/minimum-moves-to-equal-array-elements-ii/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-20
"""


from typing import List


class Solution:
    def minMoves2(self, a: List[int]) -> int:
        a.sort()
        i, j = 0, len(a)-1
        ans = 0
        while i < j:
            ans += a[j]-a[i]
            i += 1
            j -= 1
        return ans


# ----------------------------------
if __name__ == "__main__":
    # nums = [1, 2, 3]
    nums = [1, 2, 3, 4]

    solution = Solution()
    min_move_steps = solution.minMoves2(nums)
    print(f"The solution of this problem is : {min_move_steps}")

File: C++/minMoves2.cpp
/**
 * @File    : minMoves2.cpp
 * @Brief   : 最少移动次数使数组元素相等 II
 * @Link    : https://leetcode-cn.com/problems/minimum-moves-to-equal-array-elements-ii/
 * @Author  : Wei Li
 * @Date    : 2021-07-20
*/

#include <iostream>
#include <vector>
#include <algorithm>

class Solution
{
public:
    int minMoves2(std::vector<int> &nums)
    {
        int mid = static_cast<int>(nums.size()) / 2, ans = 0;
        std::nth_element(begin(nums), next(begin(nums), mid), end(nums));
        std::for_each(begin(nums), end(nums), [&](int num)
                 { ans += abs(num - nums[mid]); });

        return ans;
    }
};


template <typename T>
void DisplayElement(const T &container)
{
    std::cout << "[";
    for (auto element = container.begin(); element != container.end(); ++element)
    {
        std::cout << *element << ",";
    }
    std::cout << "]" << std::endl;
}

// -----------------------------
int main(int argc, char **argv)
{
    std::vector<int> nums{1, 2, 3};

    Solution solution;
    auto min_move_steps = solution.minMoves2(nums);
    DisplayElement(nums);
    std::cout << "The solution of this problem is " << min_move_steps << std::endl;

    return 0;
}
